src/main.rs

```diff
@@ -432,14 +438,24 @@ fn main() -> anyhow::Result<()> {
 
     touch_started_file_indicator();
 
+    thread::spawn(move || {
+        while !context.load(std::sync::atomic::Ordering::Relaxed) {
+            thread::sleep(Duration::new(0, 100000));
+        }
+
+        log::warn!("Shutting down service");
+
+        drop(toc_arc);
+        drop(settings);
+    });
```

Comment:
Doesn't look like especially elegant solution, could you please describe the motivation?

Reply:
This is much more elegant than it looks, I guess. This just loops over the context waiting for it to become true, providing plenty of opportunities for the thread scheduler to sleep the thread and handle other things.

There's not much better way of doing this, that I can think of. You have to know when the signal has fired. This is not very different from how programming languages handle signals internally after receiving the C handler.

Reply:
how about this https://github.com/qdrant/qdrant/pull/2137 ?

Reply:
that still only shuts down the grpc service? the several tokio runtimes are not given the opportunity to handle their own termination

Reply:
what do you mean exactly? There are `general_runtime` and `update_runtime`, they are dropped along with table of content. If there are running blocking tasks, they won't be killed. Async tasks won't be performed. Is there any problem here?

Reply:
> This is much more elegant than it looks, I guess. This just loops over the context waiting for it to become true, providing plenty of opportunities for the thread scheduler to sleep the thread and handle other things.

What about a [channel](https://doc.rust-lang.org/std/?search=channel), a [condvar](https://doc.rust-lang.org/std/sync/struct.Condvar.html) or even a semaphore? Or am I missing something?

Reply:
a channel would have to be repopulated with the same message for every routine supervising this AtomicBool. if you're not familiar with how a condvar is implemented, it's not far off from that while loop I already wrote.